// terra-time-input/lib/TimeInput.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _classnames = _interopRequireDefault(require("classnames"));

var _bind = _interopRequireDefault(require("classnames/bind"));

var _terraThemeContext = _interopRequireDefault(require("terra-theme-context"));

var _terraFormInput = _interopRequireDefault(require("terra-form-input"));

var _terraButtonGroup = _interopRequireDefault(require("terra-button-group"));

var _reactIntl = require("react-intl");

var KeyCode = _interopRequireWildcard(require("keycode-js"));

var _TimeUtil = _interopRequireDefault(require("./TimeUtil"));

var _TimeInputModule = _interopRequireDefault(require("./TimeInput.module.scss"));

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

var cx = _bind.default.bind(_TimeInputModule.default);

var propTypes = {
  /**
   * Whether the time input should be disabled.
   */
  disabled: _propTypes.default.bool,

  /**
   * Custom input attributes that apply to the hour, minute, and second inputs.
   */
  // eslint-disable-next-line react/forbid-prop-types
  inputAttributes: _propTypes.default.object,

  /**
   * Custom input attributes to apply to the hour input
   */
  // eslint-disable-next-line react/forbid-prop-types
  hourAttributes: _propTypes.default.object,

  /**
  * @private
  * intl object programmatically imported through injectIntl from react-intl.
  * */
  intl: _reactIntl.intlShape.isRequired,

  /**
  * Whether the input displays as Incomplete. Use when no value has been provided. _(usage note: `required` must also be set)_.
  */
  isIncomplete: _propTypes.default.bool,

  /**
  * Whether the input displays as Invalid. Use when value does not meet validation pattern.
  */
  isInvalid: _propTypes.default.bool,

  /**
  * Whether the selected meridiem displays as Invalid. Use when value does not meet validation pattern.
  */
  isInvalidMeridiem: _propTypes.default.bool,

  /**
   * Custom input attributes to apply to the minutes input
   */
  // eslint-disable-next-line react/forbid-prop-types
  minuteAttributes: _propTypes.default.object,

  /**
   * Name of the time input. The name should be unique.
   */
  name: _propTypes.default.string.isRequired,

  /**
   * A callback function to execute when the entire time input component loses focus.
   * This event does not get triggered when the focus is moved from the hour input to the minute input or meridiem because the focus is still within the main time input component.
   */
  onBlur: _propTypes.default.func,

  /**
   * A callback function to execute when either the hour or minute has been changed.
   * The first parameter is the event. The second parameter is the changed time value.
   */
  onChange: _propTypes.default.func,

  /**
   * A callback function triggered when the hour input or minute input receives focus.
   */
  onFocus: _propTypes.default.func,

  /**
   * Callback ref to pass into the input dom element.
   */
  refCallback: _propTypes.default.func,

  /**
   * Whether or not the time is required.
   */
  required: _propTypes.default.bool,

  /**
   * Custom input attributes to apply to the seconds input
   */
  // eslint-disable-next-line react/forbid-prop-types
  secondAttributes: _propTypes.default.object,

  /**
   * Whether the input for seconds should be displayed or not. If true then the second field must have a valid
   * number for the overall time to be considered valid.
   */
  showSeconds: _propTypes.default.bool,

  /**
   * An ISO 8601 string representation of the time value in the input.
   */
  value: _propTypes.default.string,

  /**
   * Type of time input to initialize. Must be `24-hour` or `12-hour`.
   * The `de`, `es-ES`, `fr-FR`, `fr`, `nl-BE`, `nl`, `pt-BR`, `pt`, `sv-SE` and `sv` locales do not use the 12-hour time notation.
   * If the `variant` prop if set to `12-hour` for one of these supported locales, the variant will be ignored and defaults to `24-hour`.
   */
  variant: _propTypes.default.oneOf([_TimeUtil.default.FORMAT_12_HOUR, _TimeUtil.default.FORMAT_24_HOUR])
};
var defaultProps = {
  disabled: false,
  inputAttributes: {},
  isIncomplete: false,
  isInvalid: false,
  isInvalidMeridiem: false,
  minuteAttributes: {},
  hourAttributes: {},
  onBlur: null,
  onChange: null,
  onFocus: undefined,
  refCallback: undefined,
  required: false,
  secondAttributes: {},
  showSeconds: false,
  value: undefined,
  variant: _TimeUtil.default.FORMAT_24_HOUR
};

var TimeInput = /*#__PURE__*/function (_React$Component) {
  (0, _inherits2.default)(TimeInput, _React$Component);

  var _super = _createSuper(TimeInput);

  function TimeInput(props) {
    var _this;

    (0, _classCallCheck2.default)(this, TimeInput);
    _this = _super.call(this, props);
    var value = _this.props.value;
    var showSeconds = _this.props.showSeconds;

    if (value && !_TimeUtil.default.validateTime(value, showSeconds)) {
      if (process.env !== 'production') {
        // eslint-disable-next-line no-console
        console.warn("An invalid time value, ".concat(value, ", has been passed to the terra-time-picker. ") + 'This value has been ignored and will not be rendered. ' + "Time values must be in ".concat(showSeconds ? 'hh:mm:ss' : 'hh:mm', " format because showSeconds is ").concat(showSeconds, "."));
      }

      value = undefined;
    }

    _this.timeInputContainer = /*#__PURE__*/_react.default.createRef();
    _this.handleHourChange = _this.handleHourChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMinuteChange = _this.handleMinuteChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleSecondChange = _this.handleSecondChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleHourInputKeyDown = _this.handleHourInputKeyDown.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMinuteInputKeyDown = _this.handleMinuteInputKeyDown.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleSecondInputKeyDown = _this.handleSecondInputKeyDown.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleFocus = _this.handleFocus.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleHourFocus = _this.handleHourFocus.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMinuteFocus = _this.handleMinuteFocus.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleSecondFocus = _this.handleSecondFocus.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleHourBlur = _this.handleHourBlur.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMinuteBlur = _this.handleMinuteBlur.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleSecondBlur = _this.handleSecondBlur.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMeridiemButtonFocus = _this.handleMeridiemButtonFocus.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMeridiemButtonBlur = _this.handleMeridiemButtonBlur.bind((0, _assertThisInitialized2.default)(_this));
    _this.handleMeridiemButtonChange = _this.handleMeridiemButtonChange.bind((0, _assertThisInitialized2.default)(_this));

    var hour = _TimeUtil.default.splitHour(value);

    var meridiem;

    if (_TimeUtil.default.getVariantFromLocale(props) === _TimeUtil.default.FORMAT_12_HOUR) {
      if (!_this.props.intl.messages['Terra.timeInput.am'] || !_this.props.intl.messages['Terra.timeInput.pm']) {
        if (process.env !== 'production') {
          // eslint-disable-next-line no-console
          console.warn('This locale only uses 24 hour clock. The ante meridiem and post meridiem will not be displayed');
        }

        _this.anteMeridiem = '';
        _this.postMeridiem = '';
      } else {
        _this.anteMeridiem = _this.props.intl.formatMessage({
          id: 'Terra.timeInput.am'
        });
        _this.postMeridiem = _this.props.intl.formatMessage({
          id: 'Terra.timeInput.pm'
        });
      }

      if (hour) {
        var parsedHour = _TimeUtil.default.parseTwelveHourTime(hour, _this.anteMeridiem, _this.postMeridiem);

        hour = parsedHour.hourString; // eslint-disable-next-line prefer-destructuring

        meridiem = parsedHour.meridiem;
      } else {
        meridiem = _this.anteMeridiem;
      }
    }

    _this.state = {
      hour: hour,
      minute: _TimeUtil.default.splitMinute(value),
      second: _TimeUtil.default.splitSecond(value),
      isFocused: false,
      meridiem: meridiem,
      hourInitialFocused: false,
      minuteInitialFocused: false,
      secondInitialFocused: false
    };
    return _this;
  }

  (0, _createClass2.default)(TimeInput, [{
    key: "componentDidUpdate",
    value: function componentDidUpdate(prevProps) {
      var variant = _TimeUtil.default.getVariantFromLocale(this.props);

      if (this.props.value === prevProps.value && variant === _TimeUtil.default.getVariantFromLocale(prevProps)) {
        return;
      }

      var hour = _TimeUtil.default.splitHour(this.props.value);

      var meridiem = this.state.meridiem;

      if (variant === _TimeUtil.default.FORMAT_12_HOUR) {
        this.anteMeridiem = this.props.intl.formatMessage({
          id: 'Terra.timeInput.am'
        });
        this.postMeridiem = this.props.intl.formatMessage({
          id: 'Terra.timeInput.pm'
        });

        if (hour) {
          var parsedHour = _TimeUtil.default.parseTwelveHourTime(hour, this.anteMeridiem, this.postMeridiem);

          hour = parsedHour.hourString; // eslint-disable-next-line prefer-destructuring

          meridiem = parsedHour.meridiem;
        }
      } // eslint-disable-next-line react/no-did-update-set-state


      this.setState({
        hour: hour,
        minute: _TimeUtil.default.splitMinute(this.props.value),
        second: _TimeUtil.default.splitSecond(this.props.value),
        meridiem: meridiem
      });
    }
  }, {
    key: "handleFocus",
    value: function handleFocus(event) {
      if (this.props.onFocus && !this.timeInputContainer.current.contains(event.relatedTarget)) {
        this.props.onFocus(event);
      }

      this.setState({
        isFocused: true
      });
    }
  }, {
    key: "handleSecondFocus",
    value: function handleSecondFocus(event) {
      this.handleFocus(event);
      this.setState({
        secondInitialFocused: true
      }); // This check is _needed_ to avoid the contextual menu on mobile devices coming up every time the focus shifts.

      if (!_TimeUtil.default.isConsideredMobileDevice()) {
        this.secondInput.setSelectionRange(0, this.secondInput.value.length);
      }
    }
  }, {
    key: "handleMinuteFocus",
    value: function handleMinuteFocus(event) {
      this.handleFocus(event);
      this.setState({
        minuteInitialFocused: true
      }); // This check is _needed_ to avoid the contextual menu on mobile device coming up every time the focus shifts.

      if (!_TimeUtil.default.isConsideredMobileDevice()) {
        this.minuteInput.setSelectionRange(0, this.minuteInput.value.length);
      }
    }
  }, {
    key: "handleHourFocus",
    value: function handleHourFocus(event) {
      this.handleFocus(event);
      this.setState({
        hourInitialFocused: true
      }); // This check is _needed_ to avoid the contextual menu on mobile device coming up every time the focus shifts.

      if (!_TimeUtil.default.isConsideredMobileDevice()) {
        this.hourInput.setSelectionRange(0, this.hourInput.value.length);
      }
    }
  }, {
    key: "handleHourBlur",
    value: function handleHourBlur(event) {
      this.handleBlur(event, _TimeUtil.default.inputType.HOUR);
      this.setState({
        hourInitialFocused: false
      });
    }
  }, {
    key: "handleMinuteBlur",
    value: function handleMinuteBlur(event) {
      this.handleBlur(event, _TimeUtil.default.inputType.MINUTE);
      this.setState({
        minuteInitialFocused: false
      });
    }
  }, {
    key: "handleSecondBlur",
    value: function handleSecondBlur(event) {
      this.handleBlur(event, _TimeUtil.default.inputType.SECOND);
      this.setState({
        secondInitialFocused: false
      });
    }
  }, {
    key: "handleMeridiemButtonBlur",
    value: function handleMeridiemButtonBlur(event) {
      this.handleBlur(event, _TimeUtil.default.inputType.MERIDIEM);
    }
  }, {
    key: "handleBlur",
    value: function handleBlur(event, type) {
      this.setState({
        isFocused: false
      });

      if (type === _TimeUtil.default.inputType.HOUR || type === _TimeUtil.default.inputType.MINUTE || type === _TimeUtil.default.inputType.SECOND) {
        var stateValue = event.target.value; // Prepend a 0 to the value when losing focus and the value is single digit.

        if (stateValue.length === 1) {
          if (_TimeUtil.default.getVariantFromLocale(this.props) === _TimeUtil.default.FORMAT_12_HOUR && type === _TimeUtil.default.inputType.HOUR && stateValue === '0') {
            stateValue = '12';
          } else {
            stateValue = '0'.concat(stateValue);
          }

          this.handleValueChange(event, type, stateValue, this.state.meridiem);
        }
      }

      if (this.props.onBlur) {
        // Modern browsers support event.relatedTarget but event.relatedTarget returns null in IE 10 / IE 11.
        // IE 11 sets document.activeElement to the next focused element before the blur event is called.
        var activeTarget = event.relatedTarget ? event.relatedTarget : document.activeElement; // Handle blur only if focus has moved out of the entire time input component.

        if (!this.timeInputContainer.current.contains(activeTarget)) {
          this.props.onBlur(event);
        }
      }
    }
  }, {
    key: "handleHourChange",
    value: function handleHourChange(event) {
      var _this2 = this;

      if (!_TimeUtil.default.validNumericInput(event.target.value)) {
        return;
      }

      var inputValue = event.target.value;
      var stateValue = this.state.hour;

      var variant = _TimeUtil.default.getVariantFromLocale(this.props);

      var maxValue = variant === _TimeUtil.default.FORMAT_12_HOUR ? 12 : 23; // Ignore the entry if the value did not change or it is invalid.
      // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have to
      // check the length of inputValue to make sure that it is less then 2.

      if (inputValue === stateValue || inputValue.length > 2 || Number(inputValue) > maxValue) {
        return;
      } // If the change made was not a deletion of a digit, then prepend '0' if the input value is a
      // single digit value between 3 and 9 for a 24 hour time, or 2 and 9 for a 12 hour clock


      if (inputValue.length >= stateValue.length) {
        var digitsToPrependZero = ['3', '4', '5', '6', '7', '8', '9'];

        if (variant === _TimeUtil.default.FORMAT_12_HOUR) {
          digitsToPrependZero.push('2');
        }

        if (digitsToPrependZero.indexOf(inputValue) > -1) {
          inputValue = '0'.concat(inputValue);
        }
      }

      if (inputValue === '00' && variant === _TimeUtil.default.FORMAT_12_HOUR) {
        inputValue = '12';
      } // Move focus to the minute input if the hour input has a valid and complete entry.


      var moveFocusOnChange = function moveFocusOnChange() {
        if (inputValue.length === 2) {
          _this2.minuteInput.focus();
        }
      };

      this.handleValueChange(event, _TimeUtil.default.inputType.HOUR, inputValue, this.state.meridiem, moveFocusOnChange);
    }
  }, {
    key: "handleMinuteChange",
    value: function handleMinuteChange(event) {
      var _this3 = this;

      if (!_TimeUtil.default.validNumericInput(event.target.value)) {
        return;
      }

      var inputValue = event.target.value;
      var stateValue = this.state.minute;
      var maxValue = 59; // Ignore the entry if the value did not change or it is invalid.
      // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have
      // to check the length of inputValue to make sure that it is less then 2.

      if (inputValue === stateValue || inputValue.length > 2 || Number(inputValue) > maxValue) {
        return;
      } // If the change made was not a deletion of a digit, then prepend '0'
      // if the input value is a single digit value between 6 and 9.


      if (inputValue.length >= stateValue.length) {
        var digitsToPrependZero = ['6', '7', '8', '9'];

        if (digitsToPrependZero.indexOf(inputValue) > -1) {
          inputValue = '0'.concat(inputValue);
        }
      }

      var moveFocusOnChange = function moveFocusOnChange() {
        // Move focus to second if second is shown and minute input has a valid and complete entry
        if (inputValue.length === 2 && _this3.props.showSeconds) {
          _this3.secondInput.focus();
        }
      };

      this.handleValueChange(event, _TimeUtil.default.inputType.MINUTE, inputValue, this.state.meridiem, moveFocusOnChange);
    }
  }, {
    key: "handleSecondChange",
    value: function handleSecondChange(event) {
      if (!_TimeUtil.default.validNumericInput(event.target.value)) {
        return;
      }

      var inputValue = event.target.value;
      var stateValue = this.state.second;
      var maxValue = 59; // Ignore the entry if the value did not change or it is invalid.
      // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have
      // to check the length of inputValue to make sure that it is less then 2.

      if (inputValue === stateValue || inputValue.length > 2 || Number(inputValue) > maxValue) {
        return;
      } // If the change made was not a deletion of a digit, then prepend '0'
      // if the input value is a single digit value between 6 and 9.


      if (inputValue.length >= stateValue.length) {
        var digitsToPrependZero = ['6', '7', '8', '9'];

        if (digitsToPrependZero.indexOf(inputValue) > -1) {
          inputValue = '0'.concat(inputValue);
        }
      }

      this.handleValueChange(event, _TimeUtil.default.inputType.SECOND, inputValue, this.state.meridiem);
    }
  }, {
    key: "handleMeridiemButtonFocus",
    value: function handleMeridiemButtonFocus(event) {
      if (this.props.onFocus && !this.timeInputContainer.current.contains(event.relatedTarget)) {
        this.props.onFocus(event);
      }
    }
    /**
     * Takes a key input from the hour input, and processes it based on the value of the keycode.
     * If the key is an up or down arrow, it increments/decrements the hour. If the right arrow
     * is pressed, it shifts focus to the minute input.
     * @param {Object} event Event object generated from the event delegation.
     */

  }, {
    key: "handleHourInputKeyDown",
    value: function handleHourInputKeyDown(event) {
      var stateValue = this.state.hour;
      var meridiem = this.state.meridiem;
      var previousStateValue = stateValue;

      var variant = _TimeUtil.default.getVariantFromLocale(this.props);

      if (event.keyCode === KeyCode.KEY_UP) {
        stateValue = _TimeUtil.default.incrementHour(stateValue, variant); // Hitting 12 when incrementing up changes the meridiem

        if (variant === _TimeUtil.default.FORMAT_12_HOUR && stateValue === '12') {
          if (meridiem === this.postMeridiem || !previousStateValue) {
            meridiem = this.anteMeridiem;
          } else {
            meridiem = this.postMeridiem;
          }
        }
      }

      if (event.keyCode === KeyCode.KEY_DOWN) {
        stateValue = _TimeUtil.default.decrementHour(stateValue, variant); // Hitting 11 when incrementing down changes the meridiem

        if (variant === _TimeUtil.default.FORMAT_12_HOUR && stateValue === '11') {
          meridiem = meridiem === this.postMeridiem ? this.anteMeridiem : this.postMeridiem;
        }
      }

      if (stateValue !== previousStateValue) {
        this.handleValueChange(event, _TimeUtil.default.inputType.HOUR, stateValue, meridiem);
      }

      if (event.keyCode === KeyCode.KEY_RIGHT) {
        this.focusMinuteFromHour(event);
      }
    }
  }, {
    key: "focusMinuteFromHour",
    value: function focusMinuteFromHour(event) {
      // If the hour is empty or the cursor is after the value, move focus to the minute input when the right arrow is pressed.
      if (this.state.hour.length === 0 || this.state.hour.length === this.hourInput.selectionEnd) {
        this.minuteInput.focus();
        this.minuteInput.setSelectionRange(0, 0);
        event.preventDefault();
      }
    }
    /**
     * Takes a key input from the minute input, and processes it based on the value of the keycode.
     * If the key is an up or down arrow, it increments/decrements the minute. If the left arrow
     * is pressed, it shifts focus to the hour input. If the right arrow is pressed, it shifts
     * focus to the meridiem input.
     * @param {Object} event Event object generated from the event delegation.
     */

  }, {
    key: "handleMinuteInputKeyDown",
    value: function handleMinuteInputKeyDown(event) {
      var stateValue = this.state.minute;
      var previousStateValue = stateValue;

      if (event.keyCode === KeyCode.KEY_UP) {
        stateValue = _TimeUtil.default.incrementMinute(stateValue);
      }

      if (event.keyCode === KeyCode.KEY_DOWN) {
        stateValue = _TimeUtil.default.decrementMinute(stateValue);
      }

      if (previousStateValue !== stateValue) {
        this.handleValueChange(event, _TimeUtil.default.inputType.MINUTE, stateValue, this.state.meridiem);
      }

      if (event.keyCode === KeyCode.KEY_LEFT || event.keyCode === KeyCode.KEY_DELETE || event.keyCode === KeyCode.KEY_BACK_SPACE) {
        this.focusHour(event);
      }

      if (event.keyCode === KeyCode.KEY_RIGHT && this.props.showSeconds) {
        this.focusSecondFromMinute(event);
      }
    }
  }, {
    key: "focusHour",
    value: function focusHour(event) {
      // If the cursor is at the left most position in the minute input, is empty or the cursor is before the value,
      // move focus to the hour input
      if (this.minuteInput.selectionEnd === 0) {
        this.hourInput.focus();

        if (this.state.hour) {
          this.hourInput.setSelectionRange(this.state.hour.length, this.state.hour.length);
          event.preventDefault();
        }
      }
    }
  }, {
    key: "focusSecondFromMinute",
    value: function focusSecondFromMinute(event) {
      // If the minute is empty or the cursor is after the value, move focus to the meridiem.
      if ((this.state.minute.length === 0 || this.state.minute.length === this.minuteInput.selectionEnd) && this.secondInput) {
        this.secondInput.focus();

        if (this.state.second) {
          this.secondInput.setSelectionRange(0, 0);
          event.preventDefault();
        }
      }
    }
    /**
     * Takes a key input from the second input, and processes it based on the value of the keycode.
     * If the key is an up or down arrow, it increments/decrements the second. If the left arrow
     * is pressed, it shifts focus to the minute input. If the right arrow is pressed, it shifts
     * focus to the meridiem input.
     * @param {Object} event Event object generated from the event delegation.
     */

  }, {
    key: "handleSecondInputKeyDown",
    value: function handleSecondInputKeyDown(event) {
      var stateValue = this.state.second;
      var previousStateValue = stateValue;

      if (event.keyCode === KeyCode.KEY_UP) {
        stateValue = _TimeUtil.default.incrementSecond(stateValue);
      }

      if (event.keyCode === KeyCode.KEY_DOWN) {
        stateValue = _TimeUtil.default.decrementSecond(stateValue);
      }

      if (previousStateValue !== stateValue) {
        this.handleValueChange(event, _TimeUtil.default.inputType.SECOND, stateValue, this.state.meridiem);
      }

      if (event.keyCode === KeyCode.KEY_LEFT || event.keyCode === KeyCode.KEY_DELETE || event.keyCode === KeyCode.KEY_BACK_SPACE) {
        this.focusMinuteFromSecond(event);
      }
    }
  }, {
    key: "focusMinuteFromSecond",
    value: function focusMinuteFromSecond(event) {
      // If the cursor is at the left most position in the second input, is empty or the cursor is before the value,
      // move focus to the minute input
      if (this.secondInput.selectionEnd === 0) {
        this.minuteInput.focus();

        if (this.state.minute) {
          this.minuteInput.setSelectionRange(this.state.minute.length, this.state.minute.length);
          event.preventDefault();
        }
      }
    }
  }, {
    key: "handleValueChange",
    value: function handleValueChange(event, type, timeValue, meridiem, moveFocusOnChange) {
      if (type === _TimeUtil.default.inputType.HOUR) {
        this.setState({
          hour: timeValue,
          meridiem: meridiem,
          hourInitialFocused: false
        }, moveFocusOnChange);
      } else if (type === _TimeUtil.default.inputType.MINUTE) {
        this.setState({
          minute: timeValue,
          minuteInitialFocused: false
        }, moveFocusOnChange);
      } else {
        this.setState({
          second: timeValue,
          secondInitialFocused: false
        }, moveFocusOnChange);
      } // Input values of length 1 indicate incomplete time, which means we cannot get a
      // reliable time so onChange isn't triggered.


      if (this.props.onChange && timeValue.length !== 1) {
        var hour = type === _TimeUtil.default.inputType.HOUR ? timeValue : this.state.hour;
        var minute = type === _TimeUtil.default.inputType.MINUTE ? timeValue : this.state.minute;
        var second = type === _TimeUtil.default.inputType.SECOND ? timeValue : this.state.second;

        if (hour === '' && minute === '' && second === '') {
          this.props.onChange(event, '');
        } else {
          this.props.onChange(event, this.formatHour(hour, meridiem).concat(':', minute).concat(this.props.showSeconds ? ':'.concat(second) : ''));
        }
      }
    }
  }, {
    key: "formatHour",
    value: function formatHour(hour, meridiem) {
      if (!hour) {
        return hour;
      }

      var tempHour = parseInt(hour, 10);

      if (_TimeUtil.default.getVariantFromLocale(this.props) === _TimeUtil.default.FORMAT_12_HOUR) {
        if (meridiem === this.postMeridiem && tempHour < 12) {
          tempHour += 12;
        } else if (meridiem === this.anteMeridiem && tempHour === 12) {
          tempHour = 0;
        }
      }

      var hourString = tempHour.toString();

      if (hourString.length < 2) {
        hourString = '0'.concat(hourString);
      }

      return hourString;
    }
  }, {
    key: "handleMeridiemButtonChange",
    value: function handleMeridiemButtonChange(event, selectedKey) {
      this.handleValueChange(event, _TimeUtil.default.inputType.HOUR, this.state.hour.toString(), selectedKey);
    }
  }, {
    key: "render",
    value: function render() {
      var _this4 = this;

      var _this$props = this.props,
          disabled = _this$props.disabled,
          inputAttributes = _this$props.inputAttributes,
          minuteAttributes = _this$props.minuteAttributes,
          hourAttributes = _this$props.hourAttributes,
          intl = _this$props.intl,
          isIncomplete = _this$props.isIncomplete,
          isInvalid = _this$props.isInvalid,
          isInvalidMeridiem = _this$props.isInvalidMeridiem,
          onBlur = _this$props.onBlur,
          onChange = _this$props.onChange,
          onFocus = _this$props.onFocus,
          name = _this$props.name,
          _refCallback = _this$props.refCallback,
          required = _this$props.required,
          secondAttributes = _this$props.secondAttributes,
          showSeconds = _this$props.showSeconds,
          value = _this$props.value,
          variant = _this$props.variant,
          customProps = (0, _objectWithoutProperties2.default)(_this$props, ["disabled", "inputAttributes", "minuteAttributes", "hourAttributes", "intl", "isIncomplete", "isInvalid", "isInvalidMeridiem", "onBlur", "onChange", "onFocus", "name", "refCallback", "required", "secondAttributes", "showSeconds", "value", "variant"]);
      var anteMeridiemClassNames = cx(['meridiem-button', {
        'is-invalid': isInvalidMeridiem && this.state.meridiem === this.anteMeridiem
      }]);
      var postMeridiemClassNames = cx(['meridiem-button', {
        'is-invalid': isInvalidMeridiem && this.state.meridiem === this.postMeridiem
      }]);

      var variantFromLocale = _TimeUtil.default.getVariantFromLocale(this.props); // Using the state of hour, minute, and second (if shown) create a time in UTC represented in ISO 8601 format.


      var timeValue = '';

      if (this.state.hour.length > 0 || this.state.minute.length > 0 || this.state.second.length > 0 && showSeconds) {
        var hour = parseInt(this.state.hour, 10);

        if (variantFromLocale === _TimeUtil.default.FORMAT_12_HOUR && this.state.meridiem === this.postMeridiem) {
          hour += 12;
        }

        timeValue = 'T'.concat(hour, ':', this.state.minute);

        if (showSeconds) {
          timeValue = timeValue.concat(':', this.state.second);
        }
      }

      var theme = this.context;
      var timeInputClassNames = (0, _classnames.default)(cx({
        disabled: disabled
      }, 'time-input', {
        'is-focused': this.state.isFocused
      }, {
        'is-invalid': isInvalid
      }, {
        'is-incomplete': isIncomplete && required && !isInvalid && !isInvalidMeridiem
      }, theme.className), customProps.className);
      var hourClassNames = cx(['time-input-hour', {
        'with-second': showSeconds
      }, {
        'initial-focus': this.state.hourInitialFocused
      }]);
      var minuteClassNames = cx(['time-input-minute', {
        'with-second': showSeconds
      }, {
        'initial-focus': this.state.minuteInitialFocused
      }]);
      var secondClassNames = cx(['time-input-second', {
        'initial-focus': this.state.secondInitialFocused
      }]);
      /* eslint-disable jsx-a11y/no-static-element-interactions */

      return /*#__PURE__*/_react.default.createElement("div", (0, _extends2.default)({}, customProps, {
        ref: this.timeInputContainer,
        className: cx('time-input-container')
      }), /*#__PURE__*/_react.default.createElement("div", {
        className: timeInputClassNames
      }, /*#__PURE__*/_react.default.createElement("input", {
        // Create a hidden input for storing the name and value attributes to use when submitting the form.
        // The data stored in the value attribute will be the visible date in the date input but in ISO 8601 format.
        type: "hidden",
        name: name,
        value: timeValue
      }), /*#__PURE__*/_react.default.createElement(_terraFormInput.default, (0, _extends2.default)({}, inputAttributes, hourAttributes, {
        "aria-label": intl.formatMessage({
          id: 'Terra.timeInput.hours'
        }),
        refCallback: function refCallback(inputRef) {
          _this4.hourInput = inputRef;
          if (_refCallback) _refCallback(inputRef);
        },
        className: hourClassNames,
        type: "text",
        value: this.state.hour,
        name: 'terra-time-hour-'.concat(name),
        placeholder: intl.formatMessage({
          id: 'Terra.timeInput.hh'
        }),
        maxLength: "2",
        onChange: this.handleHourChange,
        onKeyDown: this.handleHourInputKeyDown,
        onFocus: this.handleHourFocus,
        onBlur: this.handleHourBlur,
        size: "2",
        pattern: "\\d*",
        disabled: disabled
      })), /*#__PURE__*/_react.default.createElement("span", {
        className: cx('time-spacer')
      }, ":"), /*#__PURE__*/_react.default.createElement(_terraFormInput.default, (0, _extends2.default)({}, inputAttributes, minuteAttributes, {
        refCallback: function refCallback(inputRef) {
          _this4.minuteInput = inputRef;
        },
        "aria-label": intl.formatMessage({
          id: 'Terra.timeInput.minutes'
        }),
        className: minuteClassNames,
        type: "text",
        value: this.state.minute,
        name: 'terra-time-minute-'.concat(name),
        placeholder: intl.formatMessage({
          id: 'Terra.timeInput.mm'
        }),
        maxLength: "2",
        onChange: this.handleMinuteChange,
        onKeyDown: this.handleMinuteInputKeyDown,
        onFocus: this.handleMinuteFocus,
        onBlur: this.handleMinuteBlur,
        size: "2",
        pattern: "\\d*",
        disabled: disabled
      })), showSeconds && /*#__PURE__*/_react.default.createElement(_react.default.Fragment, null, /*#__PURE__*/_react.default.createElement("span", {
        className: cx('time-spacer')
      }, ":"), /*#__PURE__*/_react.default.createElement(_terraFormInput.default, (0, _extends2.default)({}, inputAttributes, secondAttributes, {
        refCallback: function refCallback(inputRef) {
          _this4.secondInput = inputRef;
        },
        "aria-label": intl.formatMessage({
          id: 'Terra.timeInput.seconds'
        }),
        className: secondClassNames,
        type: "text",
        value: this.state.second,
        name: 'terra-time-second-'.concat(name),
        placeholder: intl.formatMessage({
          id: 'Terra.timeInput.ss'
        }),
        maxLength: "2",
        onChange: this.handleSecondChange,
        onKeyDown: this.handleSecondInputKeyDown,
        onFocus: this.handleSecondFocus,
        onBlur: this.handleSecondBlur,
        size: "2",
        pattern: "\\d*",
        disabled: disabled
      })))), variantFromLocale === _TimeUtil.default.FORMAT_12_HOUR && /*#__PURE__*/_react.default.createElement(_terraButtonGroup.default, {
        selectedKeys: [this.state.meridiem],
        onChange: this.handleMeridiemButtonChange,
        className: cx('meridiem-button-group')
      }, /*#__PURE__*/_react.default.createElement(_terraButtonGroup.default.Button, {
        key: this.anteMeridiem,
        className: anteMeridiemClassNames,
        text: this.anteMeridiem,
        onBlur: this.handleMeridiemButtonBlur,
        onFocus: this.handleMeridiemButtonFocus,
        isDisabled: disabled
      }), /*#__PURE__*/_react.default.createElement(_terraButtonGroup.default.Button, {
        key: this.postMeridiem,
        className: postMeridiemClassNames,
        text: this.postMeridiem,
        onBlur: this.handleMeridiemButtonBlur,
        onFocus: this.handleMeridiemButtonFocus,
        isDisabled: disabled
      })));
      /* eslint-enable jsx-a11y/no-static-element-interactions */
    }
  }]);
  return TimeInput;
}(_react.default.Component);

TimeInput.propTypes = propTypes;
TimeInput.defaultProps = defaultProps;
TimeInput.contextType = _terraThemeContext.default;

var _default = (0, _reactIntl.injectIntl)(TimeInput);

exports.default = _default;
